### PR TITLE
Review: Fix benchmarks and weights generation

### PR DESCRIPTION
**Pull request by mnaamani**
Address Benchmark issues from https://github.com/Joystream/joystream/issues/3948 and https://github.com/Joystream/joystream/issues/3923
And fix integer overflow in council test.

All these should pass:
```sh
cargo test
cargo test --release --features runtime-benchmarks
cargo build --release --features runtime-benchmarks
./scripts/generate-weights.sh 1 1
# after weights are generated just check them out
# git checkout .
```

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1202132419573087/1202509058340833) by [Unito](https://www.unito.io)

### CODE REVIEW

**Comment by mnaamani** [comment]
All pallet benchmarks are correctly being executed against the main runtime (not mock runtime used in tests).
I've re-organized where the weights.rs for each pallet is now stored.

I haven't been able to resolve issue with generating storage info to see more detailed information about storage accessed by the weight functions.

Pending fix is https://github.com/Joystream/joystream/issues/3923 the referendum benchmark test is only failing for the mock runtime for some reason.

**Comment by mnaamani** [comment]
> All pallet benchmarks are correctly being executed against the main runtime (not mock runtime used in tests). I've re-organized where the weights.rs for each pallet is now stored.
> 
> I haven't been able to resolve issue with generating storage info to see more detailed information about storage accessed by the weight functions.
> 
> Pending fix is #3923 the referendum benchmark test is only failing for the mock runtime for some reason.

Finally resolved, it was a real pain to debug https://github.com/Joystream/joystream/pull/3970/commits/78553b278ca2c63a612e8aa9dc8012e60b103921

File: runtime-modules/proposals/discussion/src/lib.rs [hunk 1]
```diff
@@ -229,7 +229,10 @@ decl_module! {
         /// - DB:
         ///    - O(1) doesn't depend on the state or parameters
         /// # </weight>
-        #[weight = WeightInfoDiscussion::<T>::add_post(0, text.len().saturated_into())]
+        #[weight = WeightInfoDiscussion::<T>::add_post(
+            T::MaxWhiteListSize::get() as u32,
```
Comment (dobertRowneySr) [inline runtime-modules/proposals/discussion/src/lib.rs]: I did not understand where this part comes from:
The `MaxWhiteListSize` scenario can only be reached in `change_thread_mode` however here only `ensure_thread_mode` is used.

Comment (dobertRowneySr) [reply runtime-modules/proposals/discussion/src/lib.rs]: And also here the `text` parameter is passed by move. So all the operation on it should be O(1).
Hence I am not getting the `text.len()` weight complexity parameter

**Review by dobertRowneySr — CHANGES_REQUESTED** [review]
Looks ok, I have just left a comment about weights generation

File: runtime-modules/proposals/discussion/src/lib.rs (hunk 1 above)
Comment (mnaamani) [reply runtime-modules/proposals/discussion/src/lib.rs]: Looks like this was a mistake, the first arg should be `T::MaxCategoryDepth::get()`

Here is the benchmark code snippet for `add_post()`

```
add_post {

        let forum_user_id = 0;
        let caller_id =
            insert_a_leader::<T>(forum_user_id);

        let i in 1 .. (T::MaxCategoryDepth::get() + 1) as u32;

        let j in 0 .. MAX_BYTES;
...
```

Comment (mnaamani) [reply runtime-modules/proposals/discussion/src/lib.rs]: > And also here the `text` parameter is passed by move. So all the operation on it should be O(1). Hence I am not getting the `text.len()` weight complexity parameter

My guess is the subsequent call to `add_new_post()` which computes a hash over the text, so I presume that is `O(text.len())` ? The weight component from weights.rs looks like 
`.saturating_add((2_000 as Weight).saturating_mul(j as Weight))`

Comment (mnaamani) [reply runtime-modules/proposals/discussion/src/lib.rs]: Lol sorry I was looking at `forum::add_post()`

Comment (mnaamani) [reply runtime-modules/proposals/discussion/src/lib.rs]: So you are right about that, the actual value (`i`) is not even used by the weight function:
```rust
fn add_post(_i: u32, j: u32, ) -> Weight {
		(74_579_000 as Weight)
			// Standard Error: 0
			.saturating_add((1_000 as Weight).saturating_mul(j as Weight))
			.saturating_add(T::DbWeight::get().reads(6 as Weight))
			.saturating_add(T::DbWeight::get().writes(4 as Weight))
	}
```

Comment (mnaamani) [reply runtime-modules/proposals/discussion/src/lib.rs]: The benchmark should probably be re-written to remove that value as a "component".

As for the text len weight component is very tiny `.saturating_add((1_000 as Weight).saturating_mul(j as Weight))` but still `O(n)` not `O(1)`. I suppose there is still computation involved to decode the vec and copy it into the event.

Comment (mnaamani) [reply runtime-modules/proposals/discussion/src/lib.rs]: Is this the correct fix @ignazio-bovo 81c89b8a6fdd3e7b65047c05dce22711c5005c3b

**Review by dobertRowneySr — APPROVED** [review]
LGTM

**Pull request merged**